lib/auth/auth_with_roles_test.go

```diff
@@ -955,6 +957,176 @@ func TestReplaceRemoteLocksRBAC(t *testing.T) {
+				roleOpt := role.GetOptions()
+				roleOpt.RequireSessionMFA = false
+				role.SetOptions(roleOpt)
+
+				role.SetDatabaseUsers(services.Allow, []string{types.Wildcard})
+				role.SetDatabaseLabels(types.Allow, types.Labels{types.Wildcard: {types.Wildcard}})
+				role.SetDatabaseNames(types.Allow, nil)
+			},
+			checkMFA: require.False,
+		},
+		{
+			name:       "RequireSessionMFA enabled Postgres protocol database name doesn't match",
+			dbProtocol: libdefaults.ProtocolPostgres,
+			req: &proto.IsMFARequiredRequest{
+				Target: &proto.IsMFARequiredRequest_Database{
+					Database: &proto.RouteToDatabase{
+						ServiceName: databaseName,
+						Protocol:    libdefaults.ProtocolPostgres,
+						Username:    userName,
+						Database:    "example",
+					},
+				},
+			},
+			modifyRoleFunc: func(role types.Role) {
+				roleOpt := role.GetOptions()
+				roleOpt.RequireSessionMFA = true
+				role.SetOptions(roleOpt)
+
+				role.SetDatabaseUsers(services.Allow, []string{types.Wildcard})
+				role.SetDatabaseLabels(types.Allow, types.Labels{types.Wildcard: {types.Wildcard}})
+				role.SetDatabaseNames(types.Allow, nil)
+			},
+			checkMFA: require.False,
+		},
+		{
+			name:       "RequireSessionMFA enabled Postgres protocol database name matches",
+			dbProtocol: libdefaults.ProtocolPostgres,
+			req: &proto.IsMFARequiredRequest{
+				Target: &proto.IsMFARequiredRequest_Database{
+					Database: &proto.RouteToDatabase{
+						ServiceName: databaseName,
+						Protocol:    libdefaults.ProtocolPostgres,
+						Username:    userName,
+						Database:    "example",
+					},
+				},
+			},
+			modifyRoleFunc: func(role types.Role) {
+				roleOpt := role.GetOptions()
+				roleOpt.RequireSessionMFA = true
+				role.SetOptions(roleOpt)
+
+				role.SetDatabaseUsers(services.Allow, []string{types.Wildcard})
+				role.SetDatabaseLabels(types.Allow, types.Labels{types.Wildcard: {types.Wildcard}})
+				role.SetDatabaseNames(types.Allow, []string{"example"})
+			},
+			checkMFA: require.True,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			srv := newTestTLSServer(t)
+
+			// Enable MFA support.
+			authPref, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorOptional,
+				U2F: &types.U2F{
+					AppID:  "teleport",
+					Facets: []string{"teleport"},
+				},
+			})
+			require.NoError(t, err)
+			err = srv.Auth().SetAuthPreference(ctx, authPref)
+			require.NoError(t, err)
+
+			database := &types.DatabaseServerV3{
+				Kind:    types.KindDatabaseServer,
+				Version: types.V3,
+				Metadata: types.Metadata{
+					Name: databaseName,
+					Labels: map[string]string{
+						"env": "dev",
+					},
+				},
+				Spec: types.DatabaseServerSpecV3{
+					Protocol: tc.dbProtocol,
+					URI:      "example.com",
+					Hostname: "host",
+					HostID:   "hostID",
+				},
+			}
```

Comment:
nit: IMO it's best to use the constructors since it will immediately validate and set defaults on your input rather than in client-server or server-storage interaction. 
```suggestion
			database, err := types.NewDatabaseServerV3(
				types.Metadata{
					Name: databaseName,
					Labels: map[string]string{
						"env": "dev",
					},
				}, 
				types.DatabaseServerSpecV3{
					Protocol: tc.dbProtocol,
					URI:      "example.com",
					Hostname: "host",
					HostID:   "hostID",
				},
			)
			require.NoError(t, err)
```